src/headerssync.cpp

```diff
@@ -0,0 +1,318 @@
+    if (m_download_state == State::FINAL) return false;
+
+    // Verify that the difficulty isn't growing too fast; an adversary with
+    // limited hashing capability has a greater chance of producing a high
+    // work chain if they compress the work into as few blocks as possible,
+    // so don't let anyone give a chain that would violate the difficulty
+    // adjustment maximum.
+    if (!PermittedDifficultyTransition(m_consensus_params, current_height,
+                previous.nBits, current.nBits)) {
+        return false;
+    }
+
+    if (!CheckProofOfWork(current.GetHash(), current.nBits, m_consensus_params)) return false;
+
+    if ((current_height - m_chain_start->nHeight) % HEADER_COMMITMENT_FREQUENCY == 0) {
+        // Try to add a commitment.
+        m_header_commitments.push_back(m_hasher(current.GetHash()) & 1);
+        if (m_header_commitments.size() > m_max_commitments) {
+            // The peer's chain is too long; give up.
+            // TODO: disconnect this peer.
+            LogPrint(BCLog::NET, "headers chain is too long; giving up sync peer=%d\n", m_id);
+            return false;
+        }
+    }
+
+    m_current_chain_work += GetBlockProof(CBlockIndex(current));
+    m_last_header_received = current;
+    ++m_current_height;
+
+    return true;
+}
+
+bool HeadersSyncState::ValidateAndStoreRedownloadedHeader(const CBlockHeader& header)
+{
+    Assume(m_download_state != State::FINAL);
+    if (m_download_state == State::FINAL) return false;
+
+    // Ensure that we're working on a header that connects to the chain we're
+    // downloading.
+    if (m_redownloaded_headers.empty()) {
+        if (header.hashPrevBlock != m_chain_start->GetBlockHash()) {
+            return false;
+        }
+        m_redownload_buffer_last_height = m_chain_start->nHeight;
+    } else if (header.hashPrevBlock != m_redownload_buffer_last_hash) {
+        return false;
+    }
+
+    int64_t next_height = m_redownload_buffer_last_height + 1;
+
+    // If we're at a header for which we previously stored a commitment, verify
+    // it is correct. Failure will result in aborting download.
+    if ((next_height - m_chain_start->nHeight) % HEADER_COMMITMENT_FREQUENCY == 0) {
+         bool commitment = m_hasher(header.GetHash()) & 1;
+         if (m_header_commitments.size() == 0) {
+            // Somehow our peer managed to feed us a different chain and
+            // we've run out of commitments.
+            return false;
+        }
+        bool expected_commitment = m_header_commitments.front();
+        m_header_commitments.pop_front();
+        if (commitment != expected_commitment) {
+            return false;
+        }
+    }
+
+    // Store this header for later processing.
+    m_redownloaded_headers.push_back(header);
+    m_redownload_buffer_last_height = next_height;
+    m_redownload_buffer_last_hash = header.GetHash();
+
+    // If we're processing our target block header, which we verified has
```

Comment:
In commit "Utilize anti-DoS headers download strategy"

Could there be a concern that the last or few last headers or so get reorganized during the second phase, resulting in a mismatch at the end? If so, perhaps it's possible to instead keep track of chainwork again in the second phase rather than remembering the exact hash at which the threshold was reached in the first phase?

Reply:
I added a commit that tries to improve this.  Left it unsquashed for now so that it's easier to review; I don't believe there are any material DoS risks introduced by this change but just want to make sure we get enough eyes on this logic.